### PR TITLE
[Yara] Quick Fix for artefact recovery

### PR DESCRIPTION
**Pull request by Megafredo**
Following the discussion with @nino-filigran, we're going to go for a quick fix, which doesn't solve the problem in its entirety but does correct it on the connector. The main problem will be dealt with at a later date.

### Proposed changes

* ~(Quick Fix) Temporary solution to the artefact issue with the YARA connector. In automatic mode, when an artefact is created with a file (either through another connector or manually), the file is not immediately accessible to the YARA connector for data retrieval, leading to an error.~ `Will be resolved on another PR`

* The error when the artefact does not initially have a file has also been fixed.

* A strange behavior has also been fixed: although only one file can be associated with an artifact at the time of creation, it is possible to add others afterward. However, when manual enrichment was used, the YARA connector would only check the first file and ignore any additional ones. From now on, all added files will be taken into account during the scan.

**IMPORTANT:** Julien's directive on the real fix : (see https://github.com/OpenCTI-Platform/connectors/issues/2811#issuecomment-2437348224)
- _"Looks like we create the artifact and the upload the file. I think we need to modify the client python to use the createArtefact API containing directly the file upload."_

- _"we also need to check the artifact creation from the platform when using a standard observable instead of the Artifact."_
Details : When creating an observable of type "artifact", it is not necessary to link it to a file, unlike a full-fledged artifact which generates hashes from a file.

### Related issues

* #2546
* #2700
* #2811
* https://github.com/OpenCTI-Platform/opencti/issues/8794

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality using different use cases
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

### CODE REVIEW

**Review by Powlinett — COMMENTED** [review]
tested with malwarebazaar-recent-additions connector, it seems to work as intended ✅
code looks good to me 👍

File: internal-enrichment/yara/src/main.py [hunk 1]
```diff
@@ -17,17 +22,43 @@ def __init__(self):
...
+                 if no files are associated.
+        """
         self.helper.log_debug("Getting Artifact contents (bytes) from OpenCTI")
 
-        file_id = artifact["importFiles"][0]["id"]
-        file_url = self.octi_api_url + "/storage/get/" + file_id
-        file_content = self.helper.api.fetch_opencti_file(file_url, binary=True)
-        return file_content
+        artifact_files_contents = (
+            artifact.get("importFiles") if artifact.get("importFiles") else []
```
Comment (flavienSindou) [inline internal-enrichment/yara/src/main.py]:
```suggestion
            artifact.get("importFiles", [])
```

File: internal-enrichment/yara/src/main.py [hunk 2]
```diff
@@ -17,17 +22,43 @@ def __init__(self):
...
+                file_name = artifact_file_content.get("name")
+                file_id = artifact_file_content.get("id")
+                file_url = self.octi_api_url + "/storage/get/" + file_id
+                file_content = self.helper.api.fetch_opencti_file(file_url, binary=True)
+                files_contents.append(file_content)
+                self.helper.log_debug(
+                    f"Associated file found in Artifact with file_name :{file_name}"
+                )
+        else:
+            self.helper.log_debug("No associated files found in Artifact")
```
Comment (flavienSindou) [inline internal-enrichment/yara/src/main.py:58]:
```suggestion
            self.helper.connector_logger.debug("No associated files found in Artifact")
```

File: internal-enrichment/yara/src/main.py [hunk 3]
```diff
@@ -63,32 +94,35 @@ def _scan_artifact(self, artifact, yara_indicators) -> None:
...
-                self.helper.log_debug(
-                    f"Created Relationship from Artifact to YARA Indicator {indicator['name']}"
-                )
+        for artifact_content in artifact_contents:
+            for indicator in yara_indicators:
+                try:
+                    rule_content = indicator["pattern"]
+                    rule = yara.compile(source=rule_content)
+                except yara.SyntaxError:
+                    self.helper.log_debug(
```
Comment (flavienSindou) [inline internal-enrichment/yara/src/main.py]: I think this should at least be a log warning perhaps error

File: internal-enrichment/yara/src/main.py [hunk 4]
```diff
@@ -17,17 +22,43 @@ def __init__(self):
...
+    def _get_artifact_contents(self, artifact) -> list[bytes]:
+        """
+        Retrieves the content associated with the artefact from OpenCTI, extracts the files and downloads their
+        contents in binary format for further processing.
+
+        :param artifact: Dictionary containing all the information in the OpenCTI artefact, potentially with an
+                         ‘importFiles’ key and a list of files to be retrieved.
+        :return: List of the binary contents of the files associated with the artefact, returns an empty list `[]`
+                 if no files are associated.
+        """
```
Comment (flavienSindou) [inline internal-enrichment/yara/src/main.py:41]: Tht's helpfull ! Thanks !

**Review by flavienSindou — APPROVED** [review]
Nice overall improvement.

I have some suggestions to use full potential of dict.get method, deprecated logger usage and logging level but otherwise it looks great to me

File: internal-enrichment/yara/src/main.py (hunk 3 above)
Comment (Megafredo) [reply internal-enrichment/yara/src/main.py]: Indeed I will change that ;)

**Pull request merged**